### ISSUE TITLE
When adding dropout layer, training become slower each batch and each epoch

### ISSUE THREAD

**Issue by avielbl** [vlfeat/matconvnet#633]
Hi,

I tried this in several network architectures and by only adding one dropout layer with  rate = 0.5, the training become slower and slower until the point it is barely progress.
The weird thing is that if I stop training using ctrl+c and call cnntrain again so it will continue from last epoch, it starts from full speed again and gradually getting slower again.

I'm using Matlab 2016a with latest matconvnet ver compiled with cudnn 5.

**Comment by chrirupp** [comment]
Hi,

we experience the same problem, but we did not track it down to dropout yet. We just see that it occurs with beta20 and cudnn5 and cudnn5.1RC. With beta18 and cudnn4 it was fine. We also use Matlab2016a. OS is Windows. We are now running a test by removing dropout from our net and will report this as soon as we have results.

**Comment by avielbl** [comment]
I'm glad to see I'm not alone with this strange behavior ;)

This is very problematic as dropout is a very fundamental layer in every modern architecture.
Do you know for sure that in beta19 for example this issue wasn't exist?

**Comment by chrirupp** [comment]
No, I have not tested with beta19 at all. The newest version where I know it still works is beta18 with cudnn4, since beta18 does not compile with cudnn5 I cannot test this combination. Are you using Windows too?

**Comment by avielbl** [comment]
I'm on windows 10, matlab 2016a and matconvnet beta 20 compiled with cudnn5 on VS2013

**Comment by jstudy** [comment]
Hi, I met this problem a few days ago. I'm on windows 8, matlab 2016a and matconvnet beta 20 compiled with cudnn5.  You could make three changes like below. It solved my issue. 
`% scale = cast(1 / (1 - opts.rate), 'like', x) ;
     scale = single(1 / (1 - opts.rate));
%   mask = scale * (gpuArray.rand(size(x), classUnderlying(x)) >= opts.rate) ;
      mask = scale * (gpuArray.rand(size(x)) >= opts.rate) ;  
%    mask = scale * (rand(size(x), 'like', x) >= opts.rate) ;
       mask = scale * (rand(size(x)) >= opts.rate) ;`

**Comment by vedaldi** [comment]
Hi, thanks for pointing this out! It seems that the dropout code still triggers some obscure MATALB bug. These days we mostly use batch normalisation instead of dropout, so it went somewhat unnoticed in our tests.

We will test these changes and incorporate them as needed in the next release.

**Comment by avielbl** [comment]
Thanks jstudy for your correction. 
I understand from the corrections you performed that you suspect the casting...

I'll give it a try and see if it indeed alleviate the problem.

**Comment by jstudy** [comment]
Hi, vedaldi,
      I am still stuck in that I could not reproduce this results of matconvnet-alex  on the ILSVRC 2012 validation data (with one single centre-crop evaluation,  41.8% top-1 err. and 19.2% top- 5 err).  I tried it under 'batchNormalization' mode, using windows 8, Matlab 2016a, beta 20 and cudnn 5. The results are  44.8% top-1 err. and 21.8% top-5 err.. Can you give me some detailed suggestions.

**Comment by avielbl** [comment]
Short update: The casting corrections indeed solved the problem with my network.
Finally I can use dropout with decent training speed!

Perhaps you should suggest this fix through git so next release will be fixed?

Thank you very much!

**Comment by vedaldi** [comment]
Hi, yes, we plan to push it to git as a hot fix. Thanks for catching this.

**Comment by Goldesel23** [comment]
I tried to implement the changes sugested by @jstudy but the same decrease of speed seems to occur. The only current alternative is to go back to version 18 ?

**Comment by ZiangYan** [comment]
@jstudy thanks, works for me. My setup is Ubuntu 14.04.3, cuda 7.5, cudnn v5, commit 91399d4 (one commit newer than v1.0-beta20).

**Comment by lenck** [comment]
Yes, sorry for that, we currently do not update the binary distribution with the github bug fixes... Probably we should work on that...

**Comment by lenck** [comment]
Ok, now with the v21 release it should be fixed, I hope :)

**Comment by yushuinanrong** [comment]
@jstudy 

Hi jstudy,

Could you point out which script should add the commands you suggested?

Thanks,
Jiawei

**Comment by Goldesel23** [comment]
@yushuinanrong in vl_nndropout ;)

**Comment by xieshuaix** [comment]
I have similiar issue with finetuning vgg-m with dropout matconvnet-master cloned in October (should be newer than v21 ?), Matlab 2015b, CUDA 7.0 and cuDNN v4 under Linux. 

However training vgg-m from scratch using the Imagenet Example in MatConvNet doesn't seem to be plauged with the issue.
